### PR TITLE
Append query string + fragment to javascript: url.

### PR DESCRIPTION
**Pull request by ben0x539**
When loading a URL whose scheme is javascript, we should do what
https://html.spec.whatwg.org/multipage/browsers.html#javascript-protocol
says and append the URL's query and fragment components to the scheme
data, as well as percent- and utf-8-decode the whole thing, before
evaluating it as javascript.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8359)

### CODE REVIEW

**Comment by ben0x539** [comment]
I think this addresses #8259, but I haven't quite figured out how testing works for servo.

**Comment by eefriedman** [comment]
For testing, https://github.com/servo/servo/blob/ff2151b8bbc62fa29c90a429b7a4f12520420b6e/tests/wpt/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/015.html is an existing test for javascript URLs, and that's the directory where you would add any new tests.  (There might be other tests somewhere in the tree, but I don't know where they are off the top of my head.)

**Comment by eefriedman** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[components/script/script_task.rs, line 1658 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8359#-K2O50hTeQ-fnFZEJVNJ)** ([raw file](https://github.com/servo/servo/blob/2a48d8da8e1ef5a62e238bd61097f5c8393a9713/components/script/script_task.rs#L1658)):</sup>
It would be nice to label the various bits of code with the steps from the HTML standard.


---

<sup>**[components/script/script_task.rs, line 1659 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8359#-K2O3rsvwdT3foNAgP98)** ([raw file](https://github.com/servo/servo/blob/2a48d8da8e1ef5a62e238bd61097f5c8393a9713/components/script/script_task.rs#L1659)):</sup>
Leftover debugging code?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8359)

**Comment by ben0x539** [comment]
I can't figure out how to make a test for this.

I tried something on the basis of `<a href='javascript:void(0) ? 0 : void(a = "it works")'>...`, but it seems that servo always navigates away when such a link it clicked, even if the javascript returns void.

I tried playing around with iframes with `src='javascript:parent.a = 0 ? 0 : "it works"'` but it seems like I can't access variables of the iframe's parent.

I also tried just having an iframe with `src='javascript:0 ? 0 : "it works"'` and asking the iframe for its contents, but I can't figure out how to get that right either.

**Comment by eefriedman** [comment]
> I tried something on the basis of <a href='javascript:void(0) ? 0 : void(a = "it works")'>..., but it seems that servo always navigates away when such a link it clicked, even if the javascript returns void.

Mmm... this doesn't work because the "If the result of executing the script is void (there is no return value),"... step isn't implemented correctly.  Actually, the whole algorithm is running way too late: the javascript code is supposed to run against the existing window/document, not a new one.

> I tried playing around with iframes with src='javascript:parent.a = 0 ? 0 : "it works"' but it seems like I can't access variables of the iframe's parent.

Hmm... looks like parent doesn't work correctly for some reason.  Not sure what's going on with that.

> I also tried just having an iframe with src='javascript:0 ? 0 : "it works"' and asking the iframe for its contents, but I can't figure out how to get that right either.

Servo doesn't compute the origin for "javascript" URLs correctly. so contentDocument returns null.

----

Given all of that, it's not actually impossible to write a test: you could write a test which navigates to a javascript URL which returns a `<script>` tag which uses location.assign to navigate to some URL with the same origin as the test.  That's extremely nasty, though...

@Ms2ger Suggestions?

**Comment by eefriedman** [comment]
Thinking about it a bit more, it's probably okay to merge this, and leave the issue open to actually add tests.

Writing `&*` usually isn't necessary in Rust; just `&` is usually equivalent.

The way this decodes the URL is a bit inefficient in terms of string copying; either we don't care, and you should just unconditionally clone, or we do care, and you should use `percent_encoding::percent_decode_to` to build up a string, then `from_utf8_lossy` the resulting `Vec<u8>`.

**Comment by Ms2ger** [comment]
@jdm thoughts on testing?

**Comment by ben0x539** [comment]
@eefriedman re: string copying, is that what you had in mind?

**Comment by jdm** [comment]
@Ms2ger I am inclined to add a straightforward test that fails in Servo because of missing features, and run a manual testcase against these changes to be sure they work as intended.

**Comment by ben0x539** [comment]
Here's an attempt at a simple test that fails in Servo but makes it obvious to visual inspection that the change does the right thing. The test passes in Firefox.

I hope I did that `./mach create-wpt` thing correctly and my naming choices aren't too awkward there.

File: components/script/script_task.rs [hunk 1]
```diff
@@ -1636,10 +1636,32 @@ impl ScriptTask {
...
             let _ar = JSAutoRequest::new(self.get_cx());
-            let evalstr = incomplete.url.non_relative_scheme_data().unwrap();
+            let mut script_source_bytes = Vec::new();
+            // Start with the scheme data of the parsed URL (5.), while percent-decoding (8.)
+            percent_decode_to(incomplete.url.non_relative_scheme_data().unwrap().as_bytes(),
+                              &mut script_source_bytes);
+            // Append question mark and query component, if any (6.), while percent-decoding (8.)
+            if let Some(ref query) = incomplete.url.query {
+                    script_source_bytes.push(b'?');
+                    percent_decode_to(query.as_bytes(), &mut script_source_bytes);
```
Comment (eefriedman) [inline components/script/script_task.rs]: Weird indentation.

Otherwise looks good!

**Comment by ben0x539** [comment]
@eefriedman Whoops, sorry about that. Indentation adjusted.

**Pull request merged**